File: WebBoggler/WebBoggler/Hourglass.cs
using System;
using System.Windows.Threading;
using System.Windows;
using System.Windows.Media;
using System.Windows.Shapes;

namespace WebBoggler
{

    public class Hourglass
    {

        public class HourglassTickEventArgs : EventArgs
        {

            private TimeSpan ElapsedTime;

            public HourglassTickEventArgs(TimeSpan elapsedTime)
            {
                ElapsedTime = elapsedTime;
            }
        }

        private DispatcherTimer _asyncTimer = new DispatcherTimer();
        private const int DefaultIntervalMs = 100;// default di 100 ms

        private DateTime _startTime;
        private DateTime _pauseTime = DateTime.MinValue;
        private DateTime _lastSecondElapsed;

        private TimeSpan _oneSecondTimeSpan = new TimeSpan(0, 0, 1);

        private bool _perpetual;
        private TimeSpan _duration;
        private bool _isRunning;

        public Hourglass()
        {
            _asyncTimer.Tick += _asyncTimer_Tick;
            _asyncTimer.Interval = new TimeSpan(0,0,0,0,DefaultIntervalMs);
            _duration = new TimeSpan(0, 0, 1, 0, 0);
            // default di 1'00"
         }

        public bool Perpetual
        {
            get
            {
                return _perpetual;
            }
            set
            {
                _perpetual = value;
            }
        }

        public TimeSpan ElapsedTime
        {
            get
            {
                if (IsRunning)
                {
                    return (DateTime.Now.ToUniversalTime() - _startTime);
                }
                else if ((_pauseTime != DateTime.MinValue))
                {
                    return (_pauseTime - _startTime);
                }
                else
                {
                    return new TimeSpan(0);
                }

            }
        }

        public double ElapsedPercent
        {
            get
            {
                if ((_startTime != DateTime.MinValue ))
                {
                    return (((double)ElapsedTime.Ticks / (double)_duration.Ticks) * 100);
                }
                else
                {
                    return 0;
                }

            }
        }

        public TimeSpan Duration
        {
            get
            {
                return _duration;
            }
            set
            {
                _duration = value;
            }
        }

        public TimeSpan RemainingTime
        {
            get
            {
                return (Duration - ElapsedTime);
            }
        }

        public bool IsRunning
        {
            get
            {
                return _isRunning;
            }
            set
            {
                _isRunning = value;
            }
        }

        public void Run()
        {
            if (_pauseTime == DateTime.MinValue )
            {
                _startTime = DateTime.Now.ToUniversalTime();
            }
            else
            {
                _startTime = (DateTime.Now.ToUniversalTime() - (_pauseTime - _startTime));
                _pauseTime = DateTime.MinValue;
            }

            _asyncTimer.Start();
            _isRunning = true;

        }

        public void Pause()
        {
            _isRunning = false;
            _pauseTime = DateTime.Now.ToUniversalTime();
            _asyncTimer.Stop();
        }

        public void Reset()
        {
            _asyncTimer.Stop();
            _isRunning = false;
            _pauseTime = DateTime.MinValue;
        }

        public DateTime StartTimeUTC
        {
            get
            {
                return _startTime;
            }
            set
            {
                _startTime = value;
            }
        }

        public delegate void ElapsedTenthDelegate(object sender, HourglassTickEventArgs e);

        public event EventHandler<HourglassTickEventArgs> ElapsedTenth;

        protected void OnElapsedTenth(object sender, HourglassTickEventArgs e)
        {
             if (ElapsedTenth != null) ElapsedTenth(this, e);
            //DateTime now = DateTime.Now;                 
        }

        public delegate void ElapsedSecondDelegate(object sender, HourglassTickEventArgs e);

        public event EventHandler<HourglassTickEventArgs> ElapsedSecond;

        protected virtual void OnElapsedSecond(object sender, HourglassTickEventArgs e)
        {
            if (ElapsedSecond != null) ElapsedSecond(this, e);

        }

        public event EventHandler TimeExpired;

        protected virtual void OnTimeExpired()
        {
            if(TimeExpired != null) TimeExpired(this, EventArgs.Empty);

        }

        private void _asyncTimer_Tick(object sender, object e)
        {
            // to try a thread-unsafe call, comment out the line before, 
            //  and uncomment the line after
            // OnTick(New TickEventArgs(_Counter))
            if (((DateTime.Now.ToUniversalTime()  - _startTime) >= _duration) && !_perpetual)
            {
                //TimeExpired(this, null);
                OnTimeExpired(); //raise TimeExpired
                this.Reset();
            }
            else
            {
                // RaiseEvent ElapsedTenth(Me, ElapsedTime, New EventArgs)
                if (((DateTime.Now.ToUniversalTime() - _lastSecondElapsed) < _oneSecondTimeSpan))
                {
                    //ElapsedTenth(this, new HourglassTickEventArgs(ElapsedTime));
                    OnElapsedTenth(this, new HourglassTickEventArgs(ElapsedTime));     //raise Tick
                }
                else
                {
                    _lastSecondElapsed = DateTime.Now.ToUniversalTime();
                    //ElapsedSecond(this, new HourglassTickEventArgs(ElapsedTime));
                    OnElapsedSecond(this, new HourglassTickEventArgs(ElapsedTime));      //raise Tick    
                }

            }
        }

    }
}

File: WebBoggler/WebBogglerServer/Hourglass.vb
Imports System.Threading

Public Class Hourglass

    Public Class HourglassTickEventArgs : Inherits EventArgs
        Public ReadOnly ElapsedTime As TimeSpan

        Public Sub New(ByVal elapsedTime As TimeSpan)
            Me.ElapsedTime = elapsedTime
        End Sub
    End Class 'TickEventArgs


    Private ReadOnly _asyncTimer As New System.Threading.Timer(AddressOf AsyncTimer_Callback, Nothing, Timeout.Infinite, Timeout.Infinite)

    Private Const DefaultIntervalMs = 100 'default di 100 ms
    Private _startTime As DateTime = Nothing
    Private _pauseTime As DateTime = Nothing
    Private _lastSecondElapsed As DateTime = Nothing
    Private ReadOnly _oneSecondTimeSpan As New TimeSpan(0, 0, 1)
    Private _perpetual As Boolean
    Private _duration As TimeSpan
    Private _isRunning As Boolean


    Public Sub New()
        _asyncTimer.Change(System.Threading.Timeout.Infinite, DefaultIntervalMs)
        _duration = New TimeSpan(0, 0, 1, 0, 0) 'default di 1'00"

    End Sub

    Public Property Perpetual As Boolean
        Get
            Return _perpetual
        End Get
        Set(value As Boolean)
            _perpetual = value
        End Set
    End Property

    Public ReadOnly Property ElapsedTime() As TimeSpan
        Get
            If IsRunning Then
                Return Now - _startTime

            Else
                If _pauseTime <> Nothing Then
                    Return _pauseTime - _startTime
                Else
                    Return Nothing
                End If
            End If
        End Get
    End Property

    Public ReadOnly Property ElapsedPercent() As Byte
        Get
            If _startTime <> Nothing Then
                Return Int((ElapsedTime).Ticks / _duration.Ticks) * 100

            Else
                Return 0
            End If
        End Get
    End Property

    Public Property Duration() As TimeSpan
        Get
            Return _duration
        End Get
        Set(ByVal value As TimeSpan)
            _duration = value
        End Set
    End Property

    Public ReadOnly Property RemainingTime() As TimeSpan
        Get
            Return (Duration - ElapsedTime)
        End Get
    End Property


    Public Property IsRunning() As Boolean
        Get
            Return _isRunning
        End Get
        Set(ByVal value As Boolean)
            _isRunning = value
        End Set
    End Property

    Public Sub Run()
        If _pauseTime = Nothing Then
            _startTime = Now
        Else
            _startTime = Now - (_pauseTime - _startTime)
            _pauseTime = Nothing
        End If
        _asyncTimer.Change(0, DefaultIntervalMs)
        _isRunning = True
    End Sub

    Public Sub Pause()

        _isRunning = False
        _pauseTime = Now
        _asyncTimer.Change(Timeout.Infinite, Timeout.Infinite)
    End Sub

    Public Sub Reset()
        _isRunning = False
        _pauseTime = Nothing
        _asyncTimer.Change(Timeout.Infinite, Timeout.Infinite)
    End Sub


    Public Delegate Sub ElapsedTenthDelegate(ByVal sender As Object, ByVal e As HourglassTickEventArgs)
    Public Event ElapsedTenth As EventHandler(Of HourglassTickEventArgs)
    Protected Overridable Sub OnElapsedTenth(ByVal sender As Object, ByVal e As HourglassTickEventArgs)
        RaiseEvent ElapsedTenth(Me, e)
    End Sub

    Public Delegate Sub ElapsedSecondDelegate(ByVal sender As Object, ByVal e As HourglassTickEventArgs)
    Public Event ElapsedSecond As EventHandler(Of HourglassTickEventArgs)
    Protected Overridable Sub OnElapsedSecond(ByVal sender As Object, ByVal e As HourglassTickEventArgs)
        RaiseEvent ElapsedSecond(Me, e)
    End Sub

    Public Event TimeExpired As EventHandler
    Protected Overridable Sub OnTimeExpired()
        RaiseEvent TimeExpired(Me, EventArgs.Empty)
    End Sub

    Private Sub AsyncTimer_Callback(ByVal state As Object)
        'to try a thread-unsafe call, comment out the line before, 
        ' and uncomment the line after
        'OnTick(New TickEventArgs(_Counter))
        If ((Now - _startTime) >= _duration) And (Not _perpetual) Then
            Reset()
            RaiseEvent TimeExpired(Me, Nothing)
            'InvokeAction(AddressOf OnTimeExpired, System.EventArgs.Empty)					 'raise TimeExpired
        Else
            'RaiseEvent ElapsedTenth(Me, ElapsedTime, New EventArgs)
            If (Now - _lastSecondElapsed) < _oneSecondTimeSpan Then
                RaiseEvent ElapsedTenth(Me, New HourglassTickEventArgs(ElapsedTime))
                'InvokeAction(AddressOf OnElapsedTenth, New HourglassTickEventArgs(ElapsedTime))	 'raise Tick
            Else
                _lastSecondElapsed = Now
                RaiseEvent ElapsedSecond(Me, New HourglassTickEventArgs(ElapsedTime))
                'InvokeAction(AddressOf OnElapsedSecond, New HourglassTickEventArgs(ElapsedTime))	  'raise Tick	
            End If
        End If
    End Sub

    Protected Overrides Sub Finalize()
        _asyncTimer.Dispose()
        MyBase.Finalize()
    End Sub
End Class
